File: python/problem2.py
#!/usr/bin/python

# Rigoberto Moreno Delgado
# John Cox

import time
import Queue

state_table = []
visited = []
parent = []
symbol = []

def main():
        

        symbols = raw_input("Enter the symbols for the set (Example: 3,4,5)\n")
        symbol_lst = [int(n) for n in symbols.split(",")]
        x = raw_input("Enter a number for X\n")
        
        t0 = time.time()
        
        make_table(symbol_lst, int(x))
        breadth_first_search(symbol_lst, int(x))
        current = 0
        str_var = str(symbol[current])
        current = parent[current]
        while(current != 0):
                str_var += str(symbol[current])
                current = parent[current]
	str_var = str_var[::-1]
        print "Shortest length string accepted by the DFA is", str_var
        t1 = time.time()
        print t1-t0 


def make_table(symbol, x):
        for i in range(x):
                temp = []
                for s in symbol:
                        temp.append(delta(i, s, x))
                state_table.append(temp)

def delta(c_state, symbol, x):
        return (c_state * 10 + symbol) % x

def breadth_first_search(symbol_lst, x):
        for i in range(x):
                parent.append(-1)
                visited.append(False)
                symbol.append(-1)
                
        q = Queue.Queue()
        q.put(0)
        good = True
        while(good):
                out = q.get()
                for s in symbol_lst:
                        nxt = delta(out, s, x)
                        if not visited[nxt]:
                                visited[nxt] = True
                                parent[nxt] = out
                                symbol[nxt] = s
                                q.put(nxt)
                        if (nxt == 0):
                                good = False
                                break
                        
if __name__ == "__main__":
        main()

File: java/Problem2.java
import java.util.Vector;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Scanner;
import java.util.Arrays;
import java.lang.StringBuilder;
import java.io.IOException;

/**
 *This is a java implementation of problem 2 from project 1
 *
 */

class Problem2 {

  static Vector<Integer[]> state_table = new Vector<Integer[]>(); 
  static Vector<Integer> parent = new Vector<Integer>();
  static Vector<Integer> symbol = new Vector<Integer>();
  static Vector<Boolean> visited = new Vector<Boolean>();

  public Integer delta(int c_state, int symbol, int x) {
    return (c_state * 10 + symbol) % x;
  }


 public void make_table(Vector<Integer> symbol, int x) {
     for (int i = 0; i < x; ++i) {
	 Integer[] temp = new Integer[symbol.size()];
	 for (int j = 0; j < symbol.size(); ++j) {
	     temp[j] = delta(i, symbol.get(j), x);
	 }
	 state_table.add(temp);
     }
 }
    
    
  public void bfs(Vector<Integer> symbol_lst, int x) {
    for (int i = 0; i < x; ++i) {
      parent.add(null);
      symbol.add(null);
      visited.add(false);
    }
    LinkedList<Integer> q = new LinkedList<Integer>();
    q.add(0);
    while(true) {
      Integer out = q.removeFirst();
      for (Integer s: symbol_lst) {
        Integer next = delta(out, s, x);
        if (!visited.get(next)) {
          visited.set(next, true);
          parent.set(next, out);
          symbol.set(next, s);
          q.add(next);
        }
        if (next == 0){
          return;
        }
      }
    }
  }

  public static void main(String[] args) {
    System.out.println("Enter the symbols for the set (Example 3,4,5)");
    Vector<Integer> numbers = new Vector<Integer>();
    Scanner numScanner = new Scanner(System.in);
    String line = numScanner.nextLine();
    String[] tokens = line.split(",");
    for (String token : tokens){
      numbers.add(Integer.parseInt(token));
    }
    System.out.println("Enter a number for X");
    int x;
    x = numScanner.nextInt();
    numScanner.nextLine();
    long startTime = System.nanoTime();
    new Problem2().make_table(numbers, x);
    new Problem2().bfs(numbers, x);
    Integer current = 0;
    String str_var = Integer.toString(symbol.get(current));
    current = parent.get(current);
    while(current != 0) {
      str_var += Integer.toString(symbol.get(current));
      current = parent.get(current);
    }
    str_var = new StringBuilder(str_var).reverse().toString();
    long endTime = System.nanoTime();
    System.out.format("Shortest length string accepted by the DFA is %s\n", str_var);
    double duration = (double)(endTime - startTime);
    System.out.format("%f seconds", duration/1000000000);
  }
}
